Feature/sdl 0238 keyboard enhancements (#238)
smartdevicelink/sdl_hmi_integration_guidelines: docs/UI/SetGlobalProperties/index.md

!!! NOTE

By default `vrHelpTitle` value is set to application name.

_**Notes for HMI expected behavior:**_

1. The system shall have the ability to receive and store multiple strings from `autoCompleteList` per application.
2. When any of the keyboard layouts is being used, the system shall reference the `autoCompleteList` strings for that application.
3. As the User enters data on the keyboard, the system shall display values from `autoCompleteList` which match the entry.
4. The number of matching `autoCompleteList` strings displayed shall only be limited by the character length constraints of the HMI.
5. The system shall provide the User with the ability to select one of the displayed matching `autoCompleteList` strings without having to enter the entire string.
6. When the User selects one of the displayed matching `autoCompleteList` strings, the system shall submit that entry and not require further User input for submission.
7. HMI allows to customize special characters depending on the keyboard layout restrictions:
    * `QWERTY` - 3 customizable keys
    * `QWERTZ` - 3 customizable keys
    * `AZERTY` - 4 customizable keys

>     * `NUMERIC` - 2 customizable keys

Author: the values are removed here https://github.com/smartdevicelink/sdl_hmi_integration_guidelines/pull/238/commits/9215059fc734e6a08c6466e09e8389af0c5b72b4

2. If at any point in time, the application sends `SetGlobalProperties` RPC with **either** of `vrHelp` **or** `helpPrompt` parameters, then SDL Core shall continue with the existing behavior of forwarding such requests to HMI and SDL Core shall not delete its internal list and shall continue to update the parameter which was not provided by the application.  
3. In case mobile application sends `AddCommand` with `CommandType = Command`, SDL Core must send updated values of `vrHelp` via `SetGlobalProperties` to HMI. _(Note: `AddCommand` requests related to choice set must NOT trigger the update of `vrHelp`)_
4. In case mobile application sends _SetGlobalProperties request_ to SDL Core:   
    * SDL Core must omit the `autoCompleteText` parameter when forwarding to the HMI
        * if `autoCompleteText` is present and `autoCompleteList` is omitted, SDL Core forwards `autoCompleteList` with a single value, taken from `autoCompleteText`.

!!! MUST

> 1. Store the information and associate it with `appID`.
>     * _Note:_ Initially, the appID together with other application-related information is provided by SDL within UpdateAppList or OnAppRegistered RPCs.
> 2. When the system receives a new list of strings in `autoCompleteList` for a particular application, the system must delete the previous list and replace it with the new list for that application. 
> 3. Whenever the User activates VR or sets up the requested values for VR help layout, HMI must display the list of commands available for voice recognition. SDL Core provides the title for this list (`vrHelpTitle` parameter) and the list of commands itself (`vrHelp` parameter which is an array of `VrHelpItem`).
>     * _**Important Note:**_ If HMI-defined VR commands are accessible together with those provided by SDL Core via VR.AddCommand, HMI must:
>         * add the corresponding VR HMI-defined commands to the list of VR help items provided by SDL via UI.SetGlobalProperties
>         * display the complete list of available VR commands (SDL-defined and HMI-defined ones) when the User activates VR.
> 4. HMI displays the in-application menu for every active application upon the User's request. It must contain SDL-requested commands (UI.AddCommand) and sub-menus (UI.AddSubMenu). SDL provides the values for the name (`menuTitle` parameter) and for the icon (`menuIcon` parameter) of this in-application menu. The values for in-application menu and touchscreen keyboard are allowed by SDL Core for navigation type of application only.
> 5. In case SDL sends request with some values in `customKeys` array, HMI must:
>     * use these values to change the special characters shown in customizable keys
>     * keep default values in the remaining keys, if the array is less than or equal to customizable keys amount

Reviewer: ```suggestion
1. Store the information and associate it with `appID`.
    * _Note:_ Initially, the appID together with other application-related information is provided by SDL within UpdateAppList or OnAppRegistered RPCs.
2. When the system receives a new list of strings in `autoCompleteList` for a particular application, the system must delete the previous list and replace it with the new list for that application. 
3. Whenever the User activates <abbr title="Voice Recognition">VR</abbr> or sets up the requested values for <abbr title="Voice Recognition">VR</abbr> help layout, HMI must display the list of commands available for voice recognition. SDL Core provides the title for this list (`vrHelpTitle` parameter) and the list of commands itself (`vrHelp` parameter which is an array of `VrHelpItem`).
    * _**Important Note:**_ If HMI-defined <abbr title="Voice Recognition">VR</abbr> commands are accessible together with those provided by SDL Core via VR.AddCommand, HMI must:
        * add the corresponding <abbr title="Voice Recognition">VR</abbr> HMI-defined commands to the list of <abbr title="Voice Recognition">VR</abbr> help items provided by SDL via UI.SetGlobalProperties
        * display the complete list of available <abbr title="Voice Recognition">VR</abbr> commands (SDL-defined and HMI-defined ones) when the User activates <abbr title="Voice Recognition">VR</abbr>.
4. HMI displays the in-application menu for every active application upon the User's request. It must contain SDL-requested commands (UI.AddCommand) and sub-menus (UI.AddSubMenu). SDL provides the values for the name (`menuTitle` parameter) and for the icon (`menuIcon` parameter) of this in-application menu. The values for in-application menu and touchscreen keyboard are allowed by SDL Core for navigation type of application only.
5. In case SDL sends request with some values in `customKeys` array, HMI must:
    * use these values to change the special characters shown in customizable keys
    * keep default values in the remaining keys, if the array is less than or equal to customizable keys amount
```
Please do not remove these abbreviation tags

_Note: HMI should not duplicate special characters on keyboard._ 
  
!!!

!!! NOTE

By default `vrHelpTitle` value is set to application name.

_**Notes for HMI expected behavior:**_

1. The system shall have the ability to receive and store multiple strings from `autoCompleteList` per application.
2. When any of the keyboard layouts are being used, the system shall reference the `autoCompleteList` strings for that application.
3. As the User enters data on the keyboard, the system shall display values from `autoCompleteList` which match the entry.
4. The number of matching `autoCompleteList` strings displayed shall only be limited by the character length constraints of the HMI.
5. The system shall provide the User with the ability to select one of the displayed matching `autoCompleteList` strings without having to enter the entire string.
6. When the User selects one of the displayed matching `autoCompleteList` strings, the system shall submit that entry and not require further User input for submission.

> 7. HMI allows to customize special characters depending on the keyboard layout restrictions.

Reviewer: ```suggestion
7. The system shall allow applications to customize the special characters in each keyboard layout depending on the [KeyboardCapabilities](../../common/structs/#keyboardcapabilities) it provides.
```

### Response

> !!! MUST

Author: @jacobkeeler  please, check https://github.com/smartdevicelink/sdl_hmi_integration_guidelines/pull/238/commits/966f4540da66425f18373069a0e677f29df99604

> 1. Store the information and associate it with `appID`.
> 
> _Note: Initially, the appID together with other application-related information is provided by SDL within UpdateAppList or OnAppRegistered RPCs._
> 

Reviewer: ```suggestion
1. Store the information and associate it with `appID`.
    * _Note:_ Initially, the appID together with other application-related information is provided by SDL within UpdateAppList or OnAppRegistered RPCs.
```
This might format better. As it is now, it breaks up the numbered list into several smaller numbered lists in the dev portal

2. When the system receives a new list of strings in `autoCompleteList` for a particular application, the system must delete the previous list and replace it with the new list for that application. 
3. Whenever the User activates VR or sets up the requested values for VR help layout, HMI must display the list of commands available for voice recognition. SDL Core provides the title for this list (`vrHelpTitle` parameter) and the list of commands itself (`vrHelp` parameter which is an array of `VrHelpItem`).

> 
> _**Important Note:**_ If HMI-defined VR commands are accessible together with those provided by SDL Core via VR.AddCommand, HMI must:* 
>   * add the corresponding VR HMI-defined commands to the list of VR help items provided by SDL via UI.SetGlobalProperties
>   * display the complete list of available VR commands (SDL-defined and HMI-defined ones) when the User activates VR.
> 

Reviewer: ```suggestion
    * _**Important Note:**_ If HMI-defined VR commands are accessible together with those provided by SDL Core via VR.AddCommand, HMI must:
        * add the corresponding VR HMI-defined commands to the list of VR help items provided by SDL via UI.SetGlobalProperties
        * display the complete list of available VR commands (SDL-defined and HMI-defined ones) when the User activates VR.
```
same rendering issues as https://github.com/smartdevicelink/sdl_hmi_integration_guidelines/pull/238/files#r573219231

5. In case SDL sends request with some values in `customKeys` array, HMI must:
    * use these values to change the special characters shown in customizable keys
    * keep default values in the remaining keys, if the array is less than or equal to customizable keys amount

_Note: HMI should not duplicate special characters on keyboard._ 
  
!!!

[Result Enumeration]: ../../common/enums/#result
!!! NOTE

In case HMI does not respond to SDL Core request during SDL-default timeout (10 sec), SDL Core returns **"GENERIC_ERROR"**  result code to the corresponding mobile application request.

!!!

#### Parameters

This RPC has no additional parameter requirements
This RPC has no additional parameter requirements.

### Sequence Diagrams

|||
SetGlobalProperties for active app on HMI with <abbr title="Voice Recognition">VR</abbr> activation
SetGlobalProperties for active app on HMI with VR activation
![SetGlobalProperties](./assets/SetGlobalPropertiesActiveVRActivate.png)
|||

|||
SetGlobalProperties for active app with TTS.SetGlobalProperties_request, UI.SetGlobalProperties_request, VR.Started, VR.Stopped.   

> ![SetGlobalProperties](./assets/SetGlobalProperties_TTS_UI_VR.png)

Author: fixed in https://github.com/smartdevicelink/sdl_hmi_integration_guidelines/pull/238/commits/4258a725c9d372bf2bf464858af0042b919235f9
